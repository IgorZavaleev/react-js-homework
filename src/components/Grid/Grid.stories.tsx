import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Grid } from "./Grid";

export default {
  title: "Grid",
  component: Grid,
} as ComponentMeta<typeof Grid>;

const Template: ComponentStory<typeof Grid> = (args) => <Grid {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  field: [
    [0, 0, 0],
    [1, 1, 1],
    [0, 0, 0],
  ],
  cellSize: "big",
};
