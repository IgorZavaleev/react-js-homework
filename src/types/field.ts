export type Coords = [number, number];

export type CellProps = {
  coords: Coords;
  state: CellState;
  onClick: onClickType;
  size: GameSize;
};

export type GameSize = "small" | "medium" | "big";
export type Density = "low" | "medium" | "high";
export type GameSpeed = "pause" | "slow" | "medium" | "fast";

export type CellState = 0 | 1;

export type FieldType = CellState[][];
export type onClickType = (coords: Coords) => void;

export type GameState = {
  field: FieldType;
  onClick: onClickType;
  setGameSize: (gameSize: GameSize) => void;
  gameSize: GameSize;
  setGameSpeed: (gameSpeed: GameSpeed) => void;
  gameSpeed: GameSpeed;
  generationNumber: number;
  setDensity: (density: Density) => void;
  clearField: () => void;
};
