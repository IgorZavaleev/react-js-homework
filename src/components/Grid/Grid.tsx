import React from "react";
import { FC } from "react";
import { GameSize, FieldType } from "@/types/field";
import { Cell } from "@/components/Cell/Cell";
import { onClickType } from "@/types/field";
import styled from "@emotion/styled";

interface WrapperProps {
  size: number;
}

const RowWrapper = styled.div<WrapperProps>`
  display: flex;
  flex-direction: row;
`;

const GridWrapper = styled.div<WrapperProps>`
  display: flex;
  flex-direction: column;
`;

export type GridProps = {
  field: FieldType;
  onClick: onClickType;
  cellSize: GameSize;
};

export const Grid: FC<GridProps> = ({ field, onClick, cellSize }) => (
  <GridWrapper>
    {field.map((row, y) => (
      <RowWrapper key={`${y}`}>
        {row.map((cell, x) => (
          <Cell
            key={`${x}_${y}`}
            coords={[x, y]}
            state={cell}
            onClick={onClick}
            size={cellSize}
          />
        ))}
      </RowWrapper>
    ))}
  </GridWrapper>
);
