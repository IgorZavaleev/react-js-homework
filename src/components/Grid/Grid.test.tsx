import React from "react";
import { render } from "@testing-library/react";

import { Grid, GridProps } from "./Grid";
import { FieldType } from "@/types/field";

const MockFieldData: FieldType = [
  [0, 1, 0],
  [1, 0, 0],
  [0, 0, 1],
];

describe("Grid component check", () => {
  it("Grid renders correctly", () => {
    const props: GridProps = {
      onClick: jest.fn(),
      field: MockFieldData,
      cellSize: "small",
    };

    const { asFragment } = render(<Grid {...props}>{MockFieldData}</Grid>);

    expect(asFragment()).toMatchSnapshot();
  });
});
