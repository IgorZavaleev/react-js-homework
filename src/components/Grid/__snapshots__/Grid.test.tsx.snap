// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid component check Grid renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1iyoj2o"
  >
    <div
      class="css-o2j9ze"
    >
      <div
        class="css-1azqsv8"
        data-testid="cell_0_0_dead"
      />
      <div
        class="css-10ye2xw"
        data-testid="cell_1_0_alive"
      />
      <div
        class="css-1azqsv8"
        data-testid="cell_2_0_dead"
      />
    </div>
    <div
      class="css-o2j9ze"
    >
      <div
        class="css-10ye2xw"
        data-testid="cell_0_1_alive"
      />
      <div
        class="css-1azqsv8"
        data-testid="cell_1_1_dead"
      />
      <div
        class="css-1azqsv8"
        data-testid="cell_2_1_dead"
      />
    </div>
    <div
      class="css-o2j9ze"
    >
      <div
        class="css-1azqsv8"
        data-testid="cell_0_2_dead"
      />
      <div
        class="css-1azqsv8"
        data-testid="cell_1_2_dead"
      />
      <div
        class="css-10ye2xw"
        data-testid="cell_2_2_alive"
      />
    </div>
  </div>
</DocumentFragment>
`;
